File: backend-java/src/main/java/com/emplo/entity/SalaryStructure.java
package com.emplo.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.UuidGenerator;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "salary_structures")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class SalaryStructure {

    @Id
    @UuidGenerator
    @Column(columnDefinition = "uuid")
    private UUID id;

    @Column(name = "employee_id", nullable = false, unique = true, columnDefinition = "uuid")
    private UUID employeeId;

    @Column(name = "effective_date", nullable = false)
    private LocalDate effectiveDate;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(nullable = false, columnDefinition = "jsonb")
    private Map<String, Object> earnings;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(nullable = false, columnDefinition = "jsonb")
    private Map<String, Object> deductions;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "employer_contributions", nullable = false, columnDefinition = "jsonb")
    private Map<String, Object> employerContributions;

    @Column(name = "gross_salary", nullable = false, precision = 14, scale = 2)
    private BigDecimal grossSalary;

    @Column(name = "total_deductions", nullable = false, precision = 14, scale = 2)
    private BigDecimal totalDeductions;

    @Column(name = "net_salary", nullable = false, precision = 14, scale = 2)
    private BigDecimal netSalary;

    @Column(name = "employer_cost", nullable = false, precision = 14, scale = 2)
    private BigDecimal employerCost;

    @Column(name = "monthly_ctc", nullable = false, precision = 14, scale = 2)
    private BigDecimal monthlyCtc;

    @Column(name = "annual_ctc", nullable = false, precision = 14, scale = 2)
    private BigDecimal annualCtc;

    @Column(name = "revision_id", columnDefinition = "uuid")
    private UUID revisionId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "employee_id", insertable = false, updatable = false)
    private Employee employee;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "revision_id", insertable = false, updatable = false)
    private SalaryRevision revision;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;
}

File: frontend/src/api/salaryStructure.ts
import { api } from "./client";

export interface SalaryStructure {
  id: string;
  employee_id: string;
  effective_date: string;
  earnings: Record<string, number>;
  deductions: Record<string, number>;
  employer_contributions: Record<string, number>;
  gross_salary: number;
  total_deductions: number;
  net_salary: number;
  employer_cost: number;
  monthly_ctc: number;
  annual_ctc: number;
}

export interface SalaryTotals {
  gross_salary: number;
  total_deductions: number;
  net_salary: number;
  employer_cost: number;
  monthly_ctc: number;
  annual_ctc: number;
}

export const salaryStructureApi = {
  get: (employeeId: string) =>
    api.get<SalaryStructure | null>(`/salary-structure/${employeeId}`),

  getMy: () =>
    api.get<SalaryStructure | null>("/salary-structure/my"),

  getTemplate: () =>
    api.get<{ earnings: Record<string, number>; deductions: Record<string, number>; employer_contributions: Record<string, number> }>("/salary-structure/template/default"),

  save: (employeeId: string, data: {
    earnings: Record<string, number>;
    deductions: Record<string, number>;
    employer_contributions: Record<string, number>;
    effective_date?: string;
  }) => api.put<SalaryStructure>(`/salary-structure/${employeeId}`, data),

  calculate: (data: {
    earnings: Record<string, number>;
    deductions: Record<string, number>;
    employer_contributions: Record<string, number>;
  }) => api.post<SalaryTotals>("/salary-structure/calculate", data),
};
